examples/DivisorCurve.jl

```diff
@@ -0,0 +1,327 @@
+export AffineCurveDivisor, ProjCurveDivisor, iseffective, multiplicity,
+       divisor
+
+################################################################################
+
+abstract type CurveDivisor end
+
+################################################################################
+# The curve C is assume to be smooth and irreducible, but it is not checked.
+@doc Markdown.doc"""
+    AffineCurveDivisor(C::AffinePlaneCurve{S}, D::Dict{Point{S}, Int}) where S <: FieldElem
+
+Given a curve `C` which is assumed to be smooth and irreducible, return the divisor on the curve `C` defined by `D`.
+"""
+mutable struct AffineCurveDivisor{S <: FieldElem} <: CurveDivisor
+    C::AffinePlaneCurve{S}
+    divisor::Dict{Point{S}, Int}
+    degree::Int
+    function AffineCurveDivisor(C::AffinePlaneCurve{S}, D::Dict{Point{S}, Int}) where S <: FieldElem
+        for P in keys(D)
+            iszero(evaluate(C.eq, P.coord)) || error("The point ", P.coord, "is not on the curve.")
+            if iszero(D[P])
+                delete!(D, P)
+            end
+        end
+        new{S}(C,
+            D,
+            sum(values(D)))
+    end
+end
+
+function AffineCurveDivisor(C::AffinePlaneCurve{S}, P::Point{S}, m::Int=1) where S <: FieldElem
+    iszero(evaluate(C.eq, P.coord)) || error("The point is not on the curve.")
+    return AffineCurveDivisor(C, Dict(P => m))
+end
+
+################################################################################
+
+@doc Markdown.doc"""
+    ProjCurveDivisor(C::ProjPlaneCurve{S}, D::Dict{Oscar.Geometry.ProjSpcElem{S}, Int}) where S <: FieldElem
+
+Given a curve `C` which is assumed to be smooth and irreducible, return the divisor on the curve `C` defined by `D`.
+"""
+mutable struct ProjCurveDivisor{S <: FieldElem} <: CurveDivisor
+    C::ProjPlaneCurve{S}
+    divisor::Dict{Oscar.Geometry.ProjSpcElem{S}, Int}
+    degree::Int
+    function ProjCurveDivisor(C::ProjPlaneCurve{S}, D::Dict{Oscar.Geometry.ProjSpcElem{S}, Int}) where S <: FieldElem
+        for P in keys(D)
+            iszero(evaluate(C.eq, P.v)) || error("The point ", P, "is not on the curve.")
```

Comment:
```suggestion
            iszero(evaluate(C.eq, P.v)) || error("The point ", P, " is not on the curve")
```
The error here, in the alternate ProjCurveDivisor constructor (line 65) and in the `AffineCurveDivisor` constructors all differ a little bit, it seems? Perhaps it would be make sense to add a helper function `_check_on_curve(C, pt)` which performs contains the above line, specialized once for the affine and once for the projective case?

For that matter, don't you have `in` methods for those types of curves? I i.e. won't this work (and be more readable):
```suggestion
            P in C || error("The point ", P, " is not on the curve")
```

Reply:
I added a function ison_curve. I don't have a "in" method, because it wants then to iterate on the points of the curve...

Reply:
it only wants to iterate if you don't provide it: generically a in B iterates over B, but you can easily define
Base.in(pt, Crv) = ison_curve(pt, Crv)

Reply:
The first time I tried, I did a mistake, now it's working with defining Base.in